### PR TITLE
Check that partial quadrature functions are defined on the materials of the mesh of the grid function

### LINKED ISSUE #286: updateGridFunction dilutes interface values when the functions do not cover all the materials of the mesh
Same AI disclaimer as https://github.com/thelfer/mfem-mgis/issues/270

## Description

`makeGridFunction` / `updateGridFunction` accept a set of partial quadrature functions covering only some of the materials of the mesh of the grid function. The coefficient returns 0 in the elements without function, and `ProjectDiscCoefficient` still counts those elements in the nodal average. The values at the interface nodes are silently diluted (halved for a node shared by one covered and one uncovered element) and the uncovered region is filled with zeros. Noticed while working on #281 / #283.

`ParaviewExportIntegrationPointResultsAtNodes` and `L2Projection` avoid this by building a submesh when the functions do not cover the whole mesh, but nothing prevents a direct call with an inconsistent mesh. Corner case, from reading the code only: the export compares the selected materials with `getAssignedMaterialsIdentifiers()` rather than with the attributes of the mesh, so an attribute without behaviour integrator leads to the whole mesh being used.
  
## Checks discussed

1. the identifiers of the spaces of the functions are unique;
2. these identifiers match the attributes of the mesh of the grid function;
3. the mesh of the grid function is either the global mesh or the submesh associated with these identifiers by the mesh discretization.

`FiniteElementSpacesManager::getFiniteElementSpace(ctx, {.location, .identifiers, .number_of_components})` already implements this rule (whole mesh if all the materials are given, cached submesh otherwise), so the mesh of the grid function can be compared with the mesh of the finite element space it returns.

### PR DESCRIPTION
**Pull request by trsxvz**
Closes #286

Same AI disclaimer as https://github.com/thelfer/mfem-mgis/pull/271

## What

`makeGridFunction` / `updateGridFunction` now fail when the functions are not defined on all the materials, or on all the boundaries, of the mesh of the grid function, and on those only. Previously the elements without function returned 0 and were still counted in the nodal averages (diluted values at the interfaces, zeros elsewhere).

The check relies on `LocationIdentifier`:

1. the location identifiers of the spaces of the functions are collected (`getLocationIdentifier` on the mesh of each space, which is the main mesh for a material and the submesh of the boundary for a boundary since #291) and must be unique and of the same kind: functions defined on materials and on boundaries can not be mixed;
2. the location identifiers associated with the attributes of the mesh of the grid function are collected (`getLocationIdentifier` on that mesh, which also fails if the mesh is not managed by the discretization);
3. both sets must be equal. Since submeshes are cached per list of attributes, the mesh is then either the main mesh or the submesh returned by `getSubMesh` for these materials. A function defined on a boundary is only accepted on the submesh of this boundary, since its partial quadrature space is indexed by the elements of this submesh.

Comparing location identifiers rather than raw attributes rejects a grid function defined on the submesh of a boundary whose attribute equals the identifier of the material of the functions, and conversely. The check creates nothing (no submesh, no finite element space). Error messages use `getLocationName`.

## Projections

The projection code assumed partial quadrature spaces defined on materials, indexed by the elements of the main mesh:

  - `updateGridFunction` on a submesh always used the parent element map; for the submesh of a boundary the plain coefficients are now used, the spaces being indexed by the elements of the submesh;
  - `update` (grid function to quadrature function) and `PartialQuadratureFunction::evaluate` used the finite element space of the main mesh; they now use the one of the quadrature space, and `update` requires a grid function defined on the mesh of the space;
  - `L2Projection` only handled materials (attributes of the main mesh, submesh on materials, right-hand side integrators using the parent element map); functions defined on boundaries are projected on the submesh of the boundary;
  - `ParaviewExportIntegrationPointResultsAtNodes` built a submesh on materials from the identifiers of user-defined functions; functions defined on boundaries are now rejected with a clear message.

## Tests

`GridFunctionOnSubMeshTest` now uses TFEL/Tests (`beam-tet.mesh`, two materials, sequential and `mpirun -n 2`): a function defined on material 1 is rejected on the whole mesh, on the submesh of material 2 and on the submesh of boundary 1, and accepted on the submesh of material 1 with exact nodal values; a function defined on boundary 2 is rejected on the whole mesh, on the submesh of material 2 and when mixed with a function defined on a material, can not be updated from a grid function on the whole mesh, and is accepted on the submesh of boundary 2 with exact nodal values. `L2ProjectionTest` gets a third case projecting a function defined on a boundary of `bar.msh`. Full suite: 110/110, the boundary cases also pass on 3 and 4 ranks.

### CODE REVIEW

File: tests/GridFunctionOnSubMeshTest.cxx [hunk 1]
```diff
@@ -0,0 +1,137 @@
+/*!
+ * \file   GridFunctionOnSubMeshTest.cxx
+ * \brief  This test checks that a grid function can only be built from partial
+ * quadrature functions defined on all the materials of its mesh.
+ * \date   18/09/2026
+ */
+
```
Comment (thelfer) [inline tests/GridFunctionOnSubMeshTest.cxx:8]: Use the `TFEL/Tests` library. See various examples in the test. This allows to generate a `junit` xml file that is parsable by continuous integrations processes

File: src/PartialQuadratureFunction.cxx [hunk 2]
```diff
@@ -819,6 +820,58 @@ namespace mfem_mgis {
...
+   */
+  template <bool parallel>
+  [[nodiscard]] static bool checkGridFunctionMesh(
+      Context& ctx,
+      const std::vector<ImmutablePartialQuadratureFunctionView>& fcts,
+      const Mesh<parallel>& mesh) noexcept {
+    // locations on which the functions are defined
+    auto locations = std::set<LocationIdentifier>{};
+    for (const auto& f : fcts) {
+      const auto& qspace = f.getPartialQuadratureSpace();
```
Comment (thelfer) [inline src/PartialQuadratureFunction.cxx:855]: This test is outdated since https://github.com/thelfer/mfem-mgis/pull/291 as it assumes at a partial quadrature space is only defined on materials.

We now have to check if:

- all the quadrature functions are defined either on material or on boundaries
- check that the mesh is defined on boundaries or materials and that this is consistent with the definition of the partial quadrature functions

Error messages must be corrected.

We also must now check that the work done the projection classes is done properly, which is probably not the case

File: include/MFEMMGIS/MeshDiscretization.hxx [hunk 3]
```diff
@@ -748,7 +748,7 @@ namespace mfem_mgis {
    */
   [[nodiscard]] constexpr bool isInvalid(const LocationIdentifier& l) noexcept {
     const auto mok = isValid(l.material_identifier);
-    const auto bok = isValid(l.material_identifier);
+    const auto bok = isValid(l.boundary_identifier);
```
Comment (thelfer) [inline include/MFEMMGIS/MeshDiscretization.hxx:768]: Please fix this in a dedicated PR

File: src/MeshDiscretization.cxx [hunk 4]
```diff
@@ -852,7 +853,8 @@ namespace mfem_mgis {
       }
       const auto* const sm = dynamic_cast<const SubMesh<false>*>(&m);
       if (sm == nullptr) {
-        return ctx.registerErrorMessage("given mesh is not a submesh");
+        // the main mesh is defined on materials
+        return true;
       }
       for (const auto& [l, ptr] : this->sequential_submeshes) {
```
Comment (thelfer) [inline src/MeshDiscretization.cxx]: Please fix that in a dedicated PR

Comment (thelfer) [reply src/MeshDiscretization.cxx]: Done in https://github.com/thelfer/mfem-mgis/pull/291

File: include/MFEMMGIS/MeshDiscretization.hxx (hunk 3 above)
Comment (thelfer) [reply include/MFEMMGIS/MeshDiscretization.hxx:768]: Done in https://github.com/thelfer/mfem-mgis/pull/291

File: src/PartialQuadratureFunction.cxx (hunk 2 above)
Comment (thelfer) [reply src/PartialQuadratureFunction.cxx]: Maybe we shall just not authorized projection on boundaries and open a dedicated PR. That does not prevent from making the previous changes to prepare this future work

**Comment by trsxvz** [comment]
@thelfer applied your recommendations
